File: clojure/complex_numbers.clj
(ns complex-numbers
  (:require [clojure.math :as math]))

(defn- sq-sum [[a b]] (+ (* a a) (* b b)))

(defn real [[a b]] a)

(defn imaginary [[a b]] b)

(def abs (comp math/sqrt sq-sum))

(defn conjugate [[a b]] [a (* -1 b)])

(defn add [[a b] [c d]] [(+ a c) (+ b d)])

(defn sub [[a b] [c d]] [(- a c) (- b d)])

(defn mul [[a b] [c d]] [(- (* a c) (* b d)) (+ (* b c) (* a d))])

(defn div
  [[a b] [c d]]
  (let [t (sq-sum [c d])]
    (if (zero? t)
      [(/ a c 1.0) 0.0]
      [(/ (+ (* a c) (* b d)) t 1.0) (/ (- (* b c) (* a d)) t 1.0)])))

File: python/complex_numbers.py
import math


class ComplexNumber:

    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return (self.real, self.imaginary) == (other.real, other.imaginary)

    def __add__(self, other):
        if isinstance(other, (int, float)):
            other = ComplexNumber(other, 0)
        return ComplexNumber(self.real + other.real,
                             self.imaginary + other.imaginary)

    def __radd__(self, other):
        return self.__add__(other)

    def __mul__(self, other):
        if isinstance(other, (int, float)):
            other = ComplexNumber(other, 0)
        return ComplexNumber(
            self.real * other.real - self.imaginary * other.imaginary,
            self.imaginary * other.real + self.real * other.imaginary,
        )

    def __rmul__(self, other):
        return self.__mul__(other)

    def __sub__(self, other):
        if isinstance(other, (int, float)):
            other = ComplexNumber(other, 0)
        return ComplexNumber(self.real - other.real,
                             self.imaginary - other.imaginary)

    def __rsub__(self, other):
        if isinstance(other, (int, float)):
            other = ComplexNumber(other, 0)
        return ComplexNumber(other.real - self.real,
                             other.imaginary - self.imaginary)

    def __truediv__(self, other):
        if isinstance(other, (int, float)):
            return ComplexNumber(self.real / other, self.imaginary / other)

        d = other.real**2 + other.imaginary**2
        return ComplexNumber(
            (self.real * other.real + self.imaginary * other.imaginary) / d,
            (self.imaginary * other.real - self.real * other.imaginary) / d,
        )

    def __rtruediv__(self, other):
        d = self.real**2 + self.imaginary**2
        if isinstance(other, (int, float)):
            return ComplexNumber(other * self.real / d,
                                 -other * self.imaginary / d)

        return ComplexNumber(
            (self.real * other.real + self.imaginary * other.imaginary) / d,
            (self.real * other.imaginary - self.imaginary * other.real) / d,
        )

    def __abs__(self):
        return (self.real**2 + self.imaginary**2)**0.5

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        er = math.e**self.real
        return ComplexNumber(er * math.cos(self.imaginary),
                             er * math.sin(self.imaginary))
